File: cpp/maximum_subarray.cc
/*
 * Filename: maximum_subarray.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sun 26 Jul 2015 01:06:50 PM EDT
 */

#include <climits>
#include <iostream>

using namespace std;

class Solution {
	public:
		int maxSubArray(vector<int>& nums) {
			int maxi = nums[0];
			int locMax = nums[0];

			for (vector<int>::size_type i = 1; i < nums.size(); i++) {
				locMax = max(locMax + nums[i], nums[i]);
				maxi = max(maxi, locMax);
			}

			return maxi;
		}
};

File: java/MaximumSubarray.java
/*
 * Filename: MaximumSubarray.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Fri 15 May 2015 07:34:04 PM EDT
 */

public class Solution {
	public int maxSubArray(int[] nums) {
		if (nums == null || nums.length == 0) {
			return 0;
		}

		int max = nums[0];
		int mmax = nums[0];
		for (int i = 1; i < nums.length; i++) {
			mmax = Math.max(nums[0], mmax + nums[i]);
			max = Math.max(mmax, max);
		}

		return max;
	}
}
